fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingHeartbeat.java

```diff
@@ -0,0 +1,186 @@
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class IndexingHeartbeat {
+    // [prefix, xid] -> [indexing-type, genesis time, heartbeat time]
+    final UUID indexerId;
+    final IndexBuildProto.IndexBuildIndexingStamp.Method indexingMethod;
+    final long genesisTimeMilliseconds;
+    final long leaseLength;
+
+    public IndexingHeartbeat(final UUID indexerId, IndexBuildProto.IndexBuildIndexingStamp.Method indexingMethod, long leaseLength) {
+        this.indexerId = indexerId;
+        this.indexingMethod = indexingMethod;
+        this.leaseLength = leaseLength;
+        this.genesisTimeMilliseconds = nowMilliseconds();
+    }
+
+    public void updateHeartbeat(@Nonnull FDBRecordStore store, @Nonnull Index index) {
+        byte[] key = IndexingSubspaces.indexheartbeatSubspace(store, index, indexerId).pack();
+        byte[] value = IndexBuildProto.IndexBuildHeartbeat.newBuilder()
+                .setMethod(indexingMethod)
+                .setGenesisTimeMilliseconds(genesisTimeMilliseconds)
+                .setHeartbeatTimeMilliseconds(nowMilliseconds())
+                .build().toByteArray();
+        store.ensureContextActive().set(key, value);
+    }
+
+    public CompletableFuture<Void> checkAndUpdateHeartbeat(@Nonnull FDBRecordStore store, @Nonnull Index index) {
+        // complete exceptionally if non-mutual, other exists
+        switch (indexingMethod) {
+            case SCRUB_REPAIR:
+            case MUTUAL_BY_RECORDS:
+                updateHeartbeat(store, index);
+                return AsyncUtil.DONE;
+
+            case BY_RECORDS:
+            case MULTI_TARGET_BY_RECORDS:
+            case BY_INDEX:
+                final AsyncIterator<KeyValue> iterator = heartbeatsIterator(store, index);
+                final long now = nowMilliseconds();
+                return AsyncUtil.whileTrue(() -> iterator.onHasNext()
+                        .thenApply(hasNext -> {
+                            if (!hasNext) {
+                                return false;
+                            }
+                            final KeyValue kv = iterator.next();
+                            try {
+                                final UUID otherIndexerId = heartbeatKeyToIndexerId(store, index, kv.getKey());
+                                if (!otherIndexerId.equals(this.indexerId)) {
+                                    final IndexBuildProto.IndexBuildHeartbeat otherHeartbeat = IndexBuildProto.IndexBuildHeartbeat.parseFrom(kv.getValue());
+                                    final long age = now - otherHeartbeat.getHeartbeatTimeMilliseconds();
+                                    if (age > 0 && age < leaseLength) {
+                                        throw new SynchronizedSessionLockedException("Failed to initialize the session because of an existing session in progress")
```

Comment:
Probably reasonable to keep this exception as is, for backwards-compatibility, but worth a comment.
Or perhaps worth throwing an exception extending from `SynchronizedSessionLockedException` (and having a comment there about why)